plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -2522,17 +2528,38 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         HiveTableHandle hiveTable = (HiveTableHandle) table;
 
         List<ColumnHandle> partitionColumns = ImmutableList.copyOf(hiveTable.getPartitionColumns());
-        List<HivePartition> partitions = partitionManager.getOrLoadPartitions(metastore, new HiveIdentity(session), hiveTable);
-
-        TupleDomain<ColumnHandle> predicate = createPredicate(partitionColumns, partitions);
-
+        TupleDomain<ColumnHandle> predicate = TupleDomain.all();
         Optional<DiscretePredicates> discretePredicates = Optional.empty();
-        if (!partitionColumns.isEmpty()) {
-            // Do not create tuple domains for every partition at the same time!
-            // There can be a huge number of partitions so use an iterable so
-            // all domains do not need to be in memory at the same time.
-            Iterable<TupleDomain<ColumnHandle>> partitionDomains = Iterables.transform(partitions, (hivePartition) -> TupleDomain.fromFixedValues(hivePartition.getKeys()));
-            discretePredicates = Optional.of(new DiscretePredicates(partitionColumns, partitionDomains));
+
+        // We won't compute the predicate and discretePredicates if only partitions names are loaded.
+        if (hiveTable.getPartitionNames().isEmpty()) {
+            Optional<List<HivePartition>> partitions = Optional.empty();
+
+            // If the partitions are not loaded, try out if they can be loaded.
+            if (hiveTable.getPartitions().isEmpty()) {
+                // Since getTableProperties doesn't allow us to update ConnectorTableHandle, we might throw away this post creating TableProperties.
+                // TODO: Allow ConnectorMetadata#getTableProperties to update ConnectorTableHandle if required.
+                HivePartitionResult partitionResult = partitionManager.getPartitions(metastore, new HiveIdentity(session), table, new Constraint(hiveTable.getEnforcedConstraint()));
+                if (partitionManager.hasPartitionsLessThanThreshold(partitionResult)) {
+                    partitions = Optional.of(partitionManager.getPartitionsAsList(partitionResult));
+                }
+            }
+            else {
+                partitions = hiveTable.getPartitions();
```

Comment:
The code here would nicely yield to `Optional#or`

```java
Optional<List<HivePartition>> partitions = hiveTable.getPartitions()
        // If the partitions are not loaded, try out if they can be loaded.
        .or(() -> {
            // Since....
            HivePartitionResult partitionResult = partitionManager.getPartitions(metastore, new HiveIdentity(session), table, new Constraint(hiveTable.getEnforcedConstraint()));
            if (...) {
                return Optional.of(partitionManager.getPartitionsAsList(partitionResult));
            }
            return Optional.empty();
        });